### DISCUSSION TITLE
SearXNG 的搜索引擎设置

### DISCUSSION THREAD

**Discussion by kitaev-chen** [coaidev/coai#305] [Discussions | 交流]
请教一下 SearXNG 在 ChatNio 中的设置有什么要注意的地方吗？

Open WebUI 中有关于 “搜索结果数量” 和 “并发请求数量” 的设置，虽然他家的搜索贼慢，但给人感觉在 SearXNG 上是可以性能调优的。ChatNio 搜索结果和速度都比较不错，但相关设置新手还是没有多少 sense。比如 “搜索引擎设置” 里，可以选择开启哪些搜索引擎。那么问题来了：

1.  已选搜索引擎的数量，对搜索性能有怎样的影响？是不是开启的越多，搜索结果越慢，给到 LLM API 的 token 越多？推荐开启数量是多少？
2. ChatNio/SearXNG 对搜索结果会做排序、过滤处理吗，还是直接拿到搜索结果发送给 LLM API？如果是直接发送，那么选择搜索引擎就偏向求质量而非求全面了。
3. 有能力在 Prompt，或者 ChatNio 中，控制搜索的时间和类别偏向吗？比如需要搜索结果是一年以内的，或者主要搜学术方面的。我发现有时候搜索某个科研领域进展，尽管选择了 arxiv，实际文献链接出现的还是有限。

**Comment by zmh-program** [comment]
得看searxng并发请求各个搜索引擎的返回时长，和数量没什么直接关系。按照需求配置搜索引擎即可，一般选google, bing, ddg三大搜索引擎即可。字符数得看返回结果，coai有字符截断相关配置。
开源版只是对联网搜索的简单调用，要支持所有模型，就只能退而求其次，这点在文档也写过，只是将prompt发给searxng，不可能提供更多的cond params。这还不是主要问题，主要问题是他只会直接转发本条消息，无context和分析过程，之前是使用davincci做总结但是质量不佳因此废弃。因为当时写开源版联网搜索的时候还没有tools calling（当时的function calling仅支持单线执行，后续gpt-4-turbo等模型才支持的）。原产品开源版2024年已经没有做功能更新了，因此在现在看来做法并不是最佳做法了。重新投入资源至开源版后，后面这个联网搜索会废弃，转换为插件的形式。

**Comment by kitaev-chen** [comment]
谢谢解释！用过别家插件形式的联网搜索，感觉返回结果都比 chatnio 的差很多。我把搜索引擎列表扔进去开启联网让 LLM 分等级小结：

- nextchat：联网插件，仅少数模型支持，很多免费模型根本用不了。模型支持的情况，返回速度尚可，但出现 token 太长的问题，只能分批次问
- open webui：联网引擎，支持 searxng。太慢，平均 200s 以上。开启 20 条，10 并发，引用链接始终是 3。
- lobechat: 联网插件，仅少数模型支持，有 nextchat 支持的它不支持，它支持的 nextchat 不支持。。。 它真的批量查询所有的搜索引擎，是直接用引擎列表的引擎开启查询。。。

感觉这些联网插件后面，还是做了一些额外工作的，然而效果却未必有 searxng 的好，还受限于各个模型的情况，似乎也不是一个很好的方案。

**Reply by zmh-program** [reply]
coai 还是用 searxng，只不过转换为插件的形式。

**Reply by omega-leo** [reply]
> coai 还是用 searxng，只不过转换为插件的形式。

请问大佬如何实现 api 调用 searxng 时指定搜索引擎呢？
